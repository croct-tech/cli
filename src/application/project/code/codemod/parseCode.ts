import {Node} from '@babel/types';
import generate from '@babel/generator';
import {Codemod, CodemodOptions, ResultCode} from '@/application/project/code/codemod/codemod';
import {Language, parse} from '@/application/project/code/codemod/parser';

export type Configuration<O extends CodemodOptions> = {
    codemod: Codemod<Node, O>,
    languages: Language[],
};

export class ParseCode<O extends CodemodOptions> implements Codemod<string, O> {
    private readonly codemod: Codemod<Node, O>;

    private readonly languages: Language[];

    public constructor(configuration: Configuration<O>) {
        this.codemod = configuration.codemod;
        this.languages = configuration.languages;
    }

    public async apply(input: string, options?: O): Promise<ResultCode<string>> {
        const result = await this.codemod.apply(parse(input, this.languages), options);

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: generate(
                result.result,
                {
                    compact: false,
                    retainFunctionParens: true,
                },
            ).code,
        };
    }
}
