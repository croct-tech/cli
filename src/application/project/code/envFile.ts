import {FileSystem} from '@/application/fs/fileSystem';

export class EnvFile {
    private readonly fileSystem: FileSystem;

    private readonly path: string;

    public constructor(fileSystem: FileSystem, path: string) {
        this.fileSystem = fileSystem;
        this.path = path;
    }

    public getName(): string {
        return this.fileSystem.getBaseName(this.path);
    }

    public exists(): Promise<boolean> {
        return this.fileSystem.exists(this.path);
    }

    public async hasVariable(name: string): Promise<boolean> {
        const content = await this.read();

        if (content === '') {
            return false;
        }

        const escapedVariable = EnvFile.escapeRegex(name);
        const regex = new RegExp(`^${escapedVariable}\\s*=`, 'm');

        return regex.test(content);
    }

    public async setVariables(variables: Record<string, string>): Promise<void> {
        const content = await this.read();

        try {
            for (const [name, value] of Object.entries(variables)) {
                await this.setVariable(name, value);
            }
        } catch (error) {
            await this.write(content);

            throw error;
        }
    }

    public async setVariable(name: string, value: string): Promise<void> {
        const content = await this.read();
        const escapedName = EnvFile.escapeRegex(name);

        if (content === '') {
            return this.write(`${name}=${value}`);
        }

        const entry = `${name}=${value}`;

        const updatedContent = content.replace(
            new RegExp(
                `${escapedName}\\s*=\\s*((?!['"\`]).*$|\`(?:\\.|[^\`])*\`|'(?:\\.|[^'])*'|"(?:\\.|[^"])*")`,
                'm',
            ),
            entry,
        );

        if (updatedContent !== content || updatedContent.includes(entry)) {
            return this.write(updatedContent);
        }

        return this.write(`${content}${content.endsWith('\n') ? '' : '\n'}${name}=${value}`);
    }

    private async write(content: string): Promise<void> {
        await this.fileSystem.writeTextFile(this.path, content, {
            overwrite: true,
        });
    }

    private async read(): Promise<string> {
        if (!await this.exists()) {
            return '';
        }

        return this.fileSystem.readTextFile(this.path);
    }

    private static escapeRegex(value: string): string {
        return value.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&');
    }
}
