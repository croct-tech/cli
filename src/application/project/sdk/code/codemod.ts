export type ResultCode<T> = {
    modified: boolean,
    result: T,
};

export type CodemodOptions = Record<string, any>;

export interface Codemod<I, O extends CodemodOptions = CodemodOptions> {
    apply(input: I, options?: O): Promise<ResultCode<I>>;
}

export class CodemodError extends Error {
    public constructor(message: string) {
        super(message);

        Object.setPrototypeOf(this, CodemodError.prototype);
    }
}

export class MalformedCodeError extends CodemodError {
    public constructor(message: string) {
        super(message);

        Object.setPrototypeOf(this, MalformedCodeError.prototype);
    }
}
