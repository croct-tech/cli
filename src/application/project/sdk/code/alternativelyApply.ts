import {Codemod, CodemodOptions, ResultCode} from '@/application/project/sdk/code/codemod';

export class AlternativelyApply<T, O extends CodemodOptions> implements Codemod<T, O> {
    private readonly codemods: Array<Codemod<T, O>>;

    public constructor(codemod: Codemod<T, O>, ...codemods: Array<Codemod<T, O>>) {
        this.codemods = [codemod, ...codemods];
    }

    public async apply(input: T, options?: O): Promise<ResultCode<T>> {
        for (const codemod of this.codemods) {
            const result = await codemod.apply(input, options);

            if (result.modified) {
                return result;
            }
        }

        return {
            modified: false,
            result: input,
        };
    }
}
