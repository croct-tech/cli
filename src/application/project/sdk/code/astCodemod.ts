import type {namedTypes as Ast} from 'ast-types';
import {print} from 'recast';
import {Codemod, ResultCode} from '@/application/project/sdk/code/transformation';
import {parse} from '@/application/project/sdk/code/parser';

export class AstCodemod implements Codemod<string> {
    private readonly codemods: Array<Codemod<Ast.Node>>;

    public constructor(...transformers: Array<Codemod<Ast.Node>>) {
        this.codemods = transformers;
    }

    public apply(input: string): ResultCode<string> {
        if (this.codemods.length === 0) {
            return {
                modified: false,
                result: input,
            };
        }

        const ast = parse(input, ['jsx', 'typescript']);

        const result = this.codemods.reduce<ResultCode<Ast.Node>>(
            (next, transformer) => {
                const transformation = transformer.apply(next.result);

                return {
                    modified: transformation.modified || next.modified,
                    result: transformation.result,
                };
            },
            {
                modified: false,
                result: ast,
            },
        );

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: print(result.result).code,
        };
    }
}
