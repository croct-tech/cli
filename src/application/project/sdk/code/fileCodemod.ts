import {readFile, writeFile} from 'fs/promises';
import {Codemod, CodemodError, ResultCode} from '@/application/project/sdk/code/transformation';
import {formatCause} from '@/application/error';

export class FileCodemod implements Codemod<string> {
    private readonly codemod: Codemod<string>;

    public constructor(codemod: Codemod<string>) {
        this.codemod = codemod;
    }

    public async apply(input: string): Promise<ResultCode<string>> {
        let source = '';

        try {
            source = await readFile(input, 'utf-8');
        } catch (error) {
            if (error.code !== 'ENOENT') {
                throw new CodemodError(`Failed to read file: ${formatCause(error)}`);
            }
        }

        const result = await this.codemod.apply(source);

        if (result.modified) {
            try {
                await writeFile(input, result.result, {encoding: 'utf-8', flag: 'w'});
            } catch (error) {
                throw new CodemodError(`Failed to write file: ${formatCause(error)}`);
            }
        }

        return result;
    }
}
