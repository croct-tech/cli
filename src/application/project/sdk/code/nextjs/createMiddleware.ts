import {namedTypes as Ast} from 'ast-types/gen/namedTypes';
import {builders as builder} from 'ast-types';
import {Codemod, ResultCode} from '@/application/project/sdk/code/codemod';
import {hasReexport} from '@/application/project/sdk/code/javascript/hasReexport';

export class CreateMiddleware implements Codemod<Ast.File> {
    public apply(input: Ast.File): Promise<ResultCode<Ast.File>> {
        const {body} = input.program;
        const exports: Ast.ExportSpecifier[] = [];

        if (!CreateMiddleware.hasReexport(input, '@croct/plug-next/middleware', 'config')) {
            exports.push(
                builder.exportSpecifier.from({
                    exported: builder.identifier('config'),
                    local: builder.identifier('config'),
                }),
            );
        }

        if (!CreateMiddleware.hasReexport(input, '@croct/plug-next/middleware', 'middleware')) {
            exports.push(
                builder.exportSpecifier.from({
                    exported: builder.identifier('middleware'),
                    local: builder.identifier('middleware'),
                }),
            );
        }

        if (exports.length > 0) {
            body.push(
                builder.exportNamedDeclaration.from({
                    declaration: null,
                    source: builder.literal('@croct/plug-next/middleware'),
                    specifiers: exports,
                }),
            );
        }

        return Promise.resolve({
            modified: true,
            result: input,
        });
    }

    private static hasReexport(input: Ast.File, moduleName: string, importName: string): boolean {
        return hasReexport(input, {
            moduleName: moduleName,
            importName: importName,
        });
    }
}
