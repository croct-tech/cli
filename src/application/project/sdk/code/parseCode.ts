import type {namedTypes as Ast} from 'ast-types';
import {print} from 'recast';
import {Codemod, CodemodOptions, ResultCode} from '@/application/project/sdk/code/codemod';
import {Language, parse} from '@/application/project/sdk/code/parser';

export type Configuration<O extends CodemodOptions> = {
    codemod: Codemod<Ast.Node, O>,
    languages: Language[],
};

export class ParseCode<O extends CodemodOptions> implements Codemod<string, O> {
    private readonly codemod: Codemod<Ast.Node, O>;

    private readonly languages: Language[];

    public constructor(configuration: Configuration<O>) {
        this.codemod = configuration.codemod;
        this.languages = configuration.languages;
    }

    public async apply(input: string, options?: O): Promise<ResultCode<string>> {
        const result = await this.codemod.apply(parse(input, this.languages), options);

        if (!result.modified) {
            return {
                modified: false,
                result: input,
            };
        }

        return {
            modified: true,
            result: print(result.result).code,
        };
    }
}
