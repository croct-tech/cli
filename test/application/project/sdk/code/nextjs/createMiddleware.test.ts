import {resolve} from 'path';
import {loadFixtures} from '../fixtures';
import {ParseCode} from '@/application/project/sdk/code/parseCode';
import {CreateMiddleware} from '@/application/project/sdk/code/nextjs/createMiddleware';

describe('CreateMiddleware', () => {
    const scenarios = loadFixtures(
        resolve(__dirname, '../fixtures/nextjs-middleware-creation'),
        {},
        {},
    );

    it.each(scenarios)('should correctly transform $name', async ({name, fixture}) => {
        const transformer = new ParseCode({
            languages: ['typescript', 'jsx'],
            codemod: new CreateMiddleware(),
        });

        const output = await transformer.apply(fixture);

        expect(output.result).toMatchSnapshot(name);
    });
});
