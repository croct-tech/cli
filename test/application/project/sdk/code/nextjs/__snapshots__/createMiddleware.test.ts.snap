// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateMiddleware should correctly transform configExport.ts: configExport.ts 1`] = `
"export { config } from "@croct/plug-next/middleware";
export { middleware } from "@croct/plug-next/middleware";
"
`;

exports[`CreateMiddleware should correctly transform empty.ts: empty.ts 1`] = `"export { config, middleware } from "@croct/plug-next/middleware";"`;

exports[`CreateMiddleware should correctly transform fullExports.ts: fullExports.ts 1`] = `
"export { config, middleware } from "@croct/plug-next/middleware";
"
`;

exports[`CreateMiddleware should correctly transform middlewareExport.ts: middlewareExport.ts 1`] = `
"export { middleware } from "@croct/plug-next/middleware";
export { config } from "@croct/plug-next/middleware";
"
`;
